Add blog "Towards Encrypted Large Language Models with FHE"

RomanBredehoft opened the pull request:
Hello, we would like to add this blog about going towards encrypted Large Language Models with Fully Homomorphic Encryption, written by @jfrery and I. !

bcm-at-zama commented on _blog.yml:
```diff
@@ -2455,3 +2455,15 @@
+
+- local: encrypted-llm
+  title: "Towards Encrypted Large Language Models with FHE"
+  author: jfrery-zama
```
not possible to have the two authors here?

jfrery replied:
I don't think so. It's only single authors. I am putting Roman here.

bcm-at-zama commented on encrypted-llm.md:
```diff
@@ -0,0 +1,149 @@
+        # Convert the input to a DualArray instance
+        q_x = DualArray(
+            float_array=self.x_calib, 
+int_array=q_hidden_states, 
```
code indentation problem

jfrery replied:
Fixed thanks

bcm-at-zama commented on encrypted-llm.md:
```diff
@@ -0,0 +1,149 @@
+        # Extract the query, key and value weight and bias values using the proper indices
+        head_0_indices = [
+            list(range(i * self.n_embd, i * self.n_embd + self.head_dim)) 
+for i in range(3)
```
code indentation problem

jfrery replied:
Fixed

bcm-at-zama commented on encrypted-llm.md:
```diff
@@ -0,0 +1,149 @@
+        # Extract the queries, keys and vales
+        q_qkv = q_qkv.expand_dims(axis=1, key=f"unsqueeze_{self.layer}")
+        q_q, q_k, q_v = q_qkv.enc_split(
+3, 
```
code indentation problem (3 lines)

bcm-at-zama commented on encrypted-llm.md:
```diff
@@ -0,0 +1,149 @@
+
+Large Language Models (LLM) have recently been proven as reliable tools for improving productivity in many areas such as programming, content creation, text analysis, web search, and distance learning.
+
+## The Impact of Large Language Models on Users' Privacy
```
Titles should be have all the same capitalization convention: either All Caps of First-letter cap

bcm-at-zama commented on encrypted-llm.md:
```diff
@@ -0,0 +1,149 @@
+
+Zama believes you can get the best of both worlds: our ambition is to protect both the privacy of the user and the IP of the model. In this blog, you’ll see how to leverage the Hugging Face transformers library and have parts of these models run on encrypted data.
+
+## Fully Homomorphic Encryption (FHE) can solve LLM privacy challenges
```
eg, problem in Title case convention

jfrery replied:
Fixed thanks.

bcm-at-zama commented on encrypted-llm.md:
```diff
@@ -0,0 +1,149 @@
+
+Figure 1 shows the GPT2 architecture which has a repeating structure: a series of multi-head attention (MHA) layers applied successively. Each MHA layer projects the inputs using the model weights, computes the attention mechanism, and re-projects the output of the attention into a new tensor.
+
+In TFHE, model weights and activations are represented with integers. Nonlinear functions must be implemented with a Programmable Bootstrapping (PBS) operation. PBS implements a table lookup (TLU) operation on encrypted data while also refreshing ciphertexts to allow [arbitrary computation](https://whitepaper.zama.ai/). On the downside, the computation time of PBS dominates the one of linear oper ...
```
Make a link for TFHE? This will not be obvious at all for HF readers

jfrery replied:
Added a link to the deep dive blog.

bcm-at-zama commented on encrypted-llm.md:
```diff
@@ -0,0 +1,149 @@
+
+## Implementation of a LLM layer with FHE
+
+Next, you’ll see how to encrypt a single attention head of the multi-head attention (MHA) block. You can also find an example for the full MHA block in this [use case example](https://github.com/zama-ai/concrete-ml/tree/17779ca571d20b001caff5792eb11e76fe2c19ba/use_case_examples/llm).
```
I would give this link "example](https://github.com/zama-ai/concrete-ml/tree/17779ca571d20b001caff5792eb11e76fe2c19ba/use_case_examples/llm)." to your code in other places, which are easier for the reader to spot: like, beginning, end or both.

bcm-at-zama reviewed (commented):
Thanks, some comments

bcm-at-zama commented:
Hello @LysandreJik (CC @fxmarty). Here is the LLM blog we had talked about, some time ago. It shows how to turn some attentions in FHE, with links to code and experiments. Hope you'll like it. Cheers!

LysandreJik commented on encrypted-llm.md:
````diff
@@ -0,0 +1,149 @@
+```
+
+Running this, you will see the following print out: “Circuit compiled with 8 bit-width”. This configuration, compatible with FHE, shows the maximum bit-width necessary to perform operations in FHE.
+
````
After the entire pipeline is explained, would it be possible to show and end-to-end example and how the data sent to the LLM is encrypted?

jfrery replied:
We have added the diagram.

LysandreJik commented on encrypted-llm.md:
```diff
@@ -0,0 +1,149 @@
+
+In transformer models, the most computationally intensive operation is the attention mechanism which multiplies the queries, keys, and values. In FHE, the cost is compounded by the specificity of multiplications in the encrypted domain. Furthermore, as the sequence length increases, the number of these challenging multiplications increases quadratically.
+
+For the encrypted head, a sequence of length 6 requires 11,622 PBS operations. This is a first experiment that has not been optimized for performance. While it can run in a matter of seconds, it would require quite a lot of computing power. Fortunately, there is an exponential increase in the performance of FHE. You can expect hybrid cleartext-FHE LLMs such, as the one discussed here, to achieve  ...
```
> You can expect hybrid cleartext-FHE LLMs such

Does this mean that the data sent to the LLM is in cleartext, and encrypted in the first attention head?

LysandreJik reviewed (commented):
Looking good! Just left a few questions that would help my personal understanding :)

@fxmarty feel free to review as well if you have time

fxmarty commented on encrypted-llm.md:
```diff
@@ -0,0 +1,149 @@
+
+Despite the appeal of LLMs, privacy concerns persist surrounding user queries that are processed by these models. On one hand, leveraging the power of LLMs is desirable but, on the other hand, there is a risk of leaking sensitive information to the LLM service provider. In some areas, such as healthcare, finance, or law, this privacy risk is a showstopper.
+
+One possible solution to this conundrum is on-premise deployment, where the LLM owner would deploy their model on the client’s machine. This is however not an optimal solution, as building an LLM may cost millions of dollars ([4.6M$ for GPT3](https://lambdalabs.com/blog/demystifying-gpt-3)) and on-premise deployment runs the risk of leaking the model intellectual property (IP).
```
conundrum may be unknown to non native English speakers (as me!)

RomanBredehoft replied:
yes indeed, I've made the change !

fxmarty commented on encrypted-llm.md:
```diff
@@ -0,0 +1,149 @@
+
+## Fully Homomorphic Encryption (FHE) Can Solve LLM Privacy Challenges
+
+Zama’s solution to the challenges of LLM deployment is to use Fully Homomorphic Encryption (FHE) which enables the execution of functions on encrypted data. It is possible to achieve the goal of protecting the model owner’s IP while still maintaining the privacy of the user's data. This demo shows that an LLM model implemented in FHE maintains the quality of the original model’s predictions. To d ...
```
```suggestion
Zama’s solution to the challenges of LLM deployment is to use Fully Homomorphic Encryption (FHE) which enables the execution of functions on encrypted data. It is possible to achieve the goal of protecting the model owner’s IP while still maintaining the privacy of the user's data. This demo shows that an LLM model implemented in FHE maintains the quality of the original model’s predictions. To do this, it’s necessary to adapt the [GPT2](https://huggingface.co/gpt2) implementation from the Hugging Face [transformers library](https://github.com/huggingface/transformers), reworking sections of the inference using Concrete-Python, which enables the conversion of Python functions into their FHE equivalents.
```

fxmarty commented on encrypted-llm.md:
```diff
@@ -0,0 +1,149 @@
+
+### Applying FHE to the Hugging Face GPT2 model
+
+Building upon the transformer library from Hugging Face, rewrite the forward pass of modules that you want to encrypt, in order to include the quantized operators. Build a SingleHeadQGPT2Model instance by first loading a [GPT2LMHeadModel](https://huggingface.co/docs/transformers/model_doc/gpt2#transformers.GPT2LMHeadModel) and then manually replace the first multi-head attention module as followi ...
```
```suggestion
Building upon the transformers library from Hugging Face, rewrite the forward pass of modules that you want to encrypt, in order to include the quantized operators. Build a SingleHeadQGPT2Model instance by first loading a [GPT2LMHeadModel](https://huggingface.co/docs/transformers/model_doc/gpt2#transformers.GPT2LMHeadModel) and then manually replace the first multi-head attention module as following:
```

fxmarty replied:
```suggestion
Building upon the transformers library from Hugging Face, rewrite the forward pass of modules that you want to encrypt, in order to include the quantized operators. Build a SingleHeadQGPT2Model instance by first loading a [GPT2LMHeadModel](https://huggingface.co/docs/transformers/model_doc/gpt2#transformers.GPT2LMHeadModel) and then manually replace the first multi-head attention module as following:
```

fxmarty commented on encrypted-llm.md:
````diff
@@ -0,0 +1,149 @@
+Building upon the transformer library from Hugging Face, rewrite the forward pass of modules that you want to encrypt, in order to include the quantized operators. Build a SingleHeadQGPT2Model instance by first loading a [GPT2LMHeadModel](https://huggingface.co/docs/transformers/model_doc/gpt2#transformers.GPT2LMHeadModel) and then manually replace the first multi-head attention module as followi ...
+
+```python
+self.transformer.h[0].attn = QGPT2SingleHeadAttention(config, n_bits=n_bits)
````
It is not clear where `QGPT2SingleHeadAttention` comes from

RomanBredehoft replied:
yes indeed, I've added a link to the source code !

fxmarty commented on encrypted-llm.md:
````diff
@@ -0,0 +1,149 @@
+The forward pass is then overwritten so that the first head of the multi-head attention mechanism, including the projections made for building the query, keys and value matrices, is performed with FHE-friendly operators. 
+
+```python
+class SingleHeadAttention(QGPT2):
````
Same for `QGPT2`

RomanBredehoft replied:
same, I've added a link to the source code !

fxmarty commented on encrypted-llm.md:
```diff
@@ -0,0 +1,149 @@
+
+In transformer models, the most computationally intensive operation is the attention mechanism which multiplies the queries, keys, and values. In FHE, the cost is compounded by the specificity of multiplications in the encrypted domain. Furthermore, as the sequence length increases, the number of these challenging multiplications increases quadratically.
+
+For the encrypted head, a sequence of length 6 requires 11,622 PBS operations. This is a first experiment that has not been optimized for performance. While it can run in a matter of seconds, it would require quite a lot of computing power. Fortunately, there is an exponential increase in the performance of FHE. You can expect hybrid cleartext-FHE LLMs such, as the one discussed here, to achieve  ...
```
Without clicking on the other blog post, it is not clear to me what

> Fortunately, there is an exponential increase in the performance of FHE.

means?

Would it be meaningful to have latency comparison with/without FHE currently? Or maybe mentioning which hardware is well suited for FHE (are GPUs fine?).

RomanBredehoft replied:
you are right, the sentence is not very clear, we'll add some additional info !

jfrery replied:
This has been rephrased

fxmarty commented on encrypted-llm.md:
```diff
@@ -0,0 +1,149 @@
+
+Large Language Models are great assistance tools in a wide variety of use cases but their implementation raises major issues for user privacy. In this blog, you saw a first step toward having the whole LLM work on encrypted data where the model would run entirely in the cloud while users' privacy would be fully respected.
+
+This step includes conversion of a specific part in a model like GPT2 to the FHE realm. This implementation leverages the transformer library from Hugging Face and allows you to evaluate the impact on the accuracy when part of the model runs on encrypted data. In addition to preserving user privacy, this approach also allows a model owner to keep a major part of their model private.
```
```suggestion
This step includes conversion of a specific part in a model like GPT2 to the FHE realm. This implementation leverages the transformers library from Hugging Face and allows you to evaluate the impact on the accuracy when part of the model runs on encrypted data. In addition to preserving user privacy, this approach also allows a model owner to keep a major part of their model private.
```

fxmarty reviewed (commented):
Looks cool! I left a few comments and leave @osanseviero to approve or not!

jfrery commented:
Hello @fxmarty @osanseviero @LysandreJik we have made changes upon you review. Let us know if you see anything else.

osanseviero commented on encrypted-llm.md:
```diff
@@ -0,0 +1,153 @@
+
+Large Language Models are great assistance tools in a wide variety of use cases but their implementation raises major issues for user privacy. In this blog, you saw a first step toward having the whole LLM work on encrypted data where the model would run entirely in the cloud while users' privacy would be fully respected.
+
+This step includes conversion of a specific part in a model like GPT2 to the FHE realm. This implementation leverages the transformers library from Hugging Face and allows you to evaluate the impact on the accuracy when part of the model runs on encrypted data. In addition to preserving user privacy, this approach also allows a model owner to keep a major part of their model private. The complete ...
```
```suggestion
This step includes the conversion of a specific part in a model like GPT2 to the FHE realm. This implementation leverages the transformers library and allows you to evaluate the impact on the accuracy when part of the model runs on encrypted data. In addition to preserving user privacy, this approach also allows a model owner to keep a major part of their model private. The complete code can be found in this [use case example](https://github.com/zama-ai/concrete-ml/tree/17779ca571d20b001caff5792eb11e76fe2c19ba/use_case_examples/llm).
```

osanseviero reviewed (commented):
Cool stuff, thanks a lot! :fire:

RomanBredehoft commented on encrypted-llm.md:
```diff
@@ -0,0 +1,153 @@
+---
```
ah yes sure ! You can find the PR here : https://huggingface.co/datasets/huggingface/documentation-images/discussions/139

RomanBredehoft commented:
Hello @fxmarty @osanseviero @LysandreJik, we addressed your latest review and moved our images to the documentation repository (the associated PR can be found here: https://huggingface.co/datasets/huggingface/documentation-images/discussions/139) ! Please let us know if we need to update anything else 🙂

RomanBredehoft commented:
@osanseviero I've just updated the date in `_blog.yml` but should we maybe agree on a specific date in advanced for this blog's publication ? Thanks !

osanseviero commented:
WDYT of publishing this wednesday next week?

RomanBredehoft commented:
Yes that's great, let's do that ! I'll update this PR accordingly next week then 🙂

osanseviero commented:
Hi there! There are some conflicts. As soon as they are resolved, I'll merge 🔥 🚀

RomanBredehoft commented:
Hello @osanseviero, conflicts are resolved 😉

osanseviero commented:
It's out! https://huggingface.co/blog/encrypted-llm

RomanBredehoft commented:
Great, thanks a lot @osanseviero !
